### PR TITLE
#27 [Inspection] Observer declaration check for duplicates

### PR DESCRIPTION
**Pull request by vasilii-b**
# Description
This PR adds a new code inspection to avoid accidentally overriding an observer.

image

## This PR solves next issues
1. magento/magento2-phpstorm-plugin#27 - Inspection for possible observer name duplication

### More details

ℹ️ The Observer Inspections adds warnings with the following information

**1. Observer name already used in the same file**
same-file-duplicate-found

**2. Observer name already used in the scope of other modules**
(see screenshot from #3)

**3. Information about the duplicated name scope: `Vendor_Module/etc`, `Vendor_Module/etc/adminhtml` or `Vendor_Module/etc/frontend`**
 
diplicated-name-hints-with-scope

**4. 🙅‍♂ No warning for the disabled observer**
disabled-observer-no-warning

**5. 📖Details of why such situation occurs and how to handle it**
inspection-description

### CODE REVIEW

File: src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                    if (!indexedEventAttributeValue.equals(eventNameAttributeValue)) {
+                        continue;
+                    }
+
+                    String indexedFileDirectory = indexedAttributeParent.getContainingDirectory().toString();
+                    String indexedFileFullPath = indexedFileDirectory.concat("/").concat(indexedAttributeParent.getName());
+                    if (indexedFileFullPath.equals(currentFileFullPath)) {
+                        continue;
+                    }
+
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java]: Hi @vasilii-b 
Could you please add all the areas (corn, webapis, etc.)?
https://devdocs.magento.com/guides/v2.3/architecture/archi_perspectives/components/modules/mod_and_areas.html

Thank you!

File: resources/inspectionDescriptions/ObserverDeclarationInspection.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<html>
```
Comment (VitaliyBoyko) [inline resources/inspectionDescriptions/ObserverDeclarationInspection.html]: Please add copyrights.

File: src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
+package com.magento.idea.magento2plugin.inspections.xml;
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java]: Copyrights.

File: src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+public class ObserverDeclarationInspection extends PhpInspection {
+
+    @NotNull
+    @Override
+    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder problemsHolder, boolean b) {
+        return new XmlElementVisitor() {
+            private final String moduleXmlFileName = ModuleXml.getInstance().getFileName();
+            private static final String eventsXmlFileName = "events.xml";//todo: move to separate file as done to ModuleXml
+            private static final String duplicatedObserverNameSameFileProblemDescription = "Observer name already used in this file. For more details see Inspection Description.";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java]: The Observer name already used in this file. For more details see Inspection Description.

File: src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            private static final String duplicatedObserverNameProblemDescription =
+                    "Observer name \"%s\" for event \"%s\" is already used in the module \"%s\" (%s scope). For more details see Inspection Description.";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java]: The Observer name...

**Comment by VitaliyBoyko** [comment]
One more thing. Don't forget to update CHANGELOG.md with the new functionality.

**Comment by vasilii-b** [comment]
Hi @VitaliyBoyko,
The requested changes have been added ✅ 
Please let me know should there be needed more adjustments.
Thanks!

File: src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                    }
+
+                    String indexedFileDirectory = indexedAttributeParent.getContainingDirectory().toString();
+                    String indexedFileFullPath = indexedFileDirectory.concat("/").concat(indexedAttributeParent.getName());
+                    if (indexedFileFullPath.equals(currentFileFullPath)) {
+                        continue;
+                    }
+
+                    if (indexedFileDirectory.contains(MagentoPackages.AREA_ADMINHTML)) {
+                        scope = MagentoPackages.AREA_ADMINHTML;
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/inspections/xml/ObserverDeclarationInspection.java]: Sorry, still did not get, why you are using here only 2 areas?

**Comment by vasilii-b** [comment]
Hi @VitaliyBoyko,
My bad here. I didn't get your message initially.
All possible areas are added now. Example:
all-possible-area-scopes-added
Please let me know if I missed smth this time.

**Pull request merged**